src/composeappmanager.cc

```diff
@@ -475,23 +475,50 @@ data::InstallationResult ComposeAppManager::finalizeInstall(const Uptane::Target
     }
     // "finalize" (run) Apps that were pulled and created before reboot
     for (const auto& app_pair : getApps(target)) {
+      //Here the installation of the apps should be checked, and if they are not installed, they should be
       const AppEngine::Result run_res = app_engine_->run({app_pair.first, app_pair.second});
       if (!run_res) {
-        const std::string err_desc{boost::str(
+        // Multiple options possible but exception is captured in general:
+        // - Possible that apps is not present due to no space available
+        // - Possible that app is not starting due to ImagePullFailure
+        // - Possible that app is not present due to new app and not pulled before update
+        // - Possible that app is not starting due to unknown reasons
+        std::string err_desc{boost::str(
             boost::format("failed to start App after booting on a new sysroot version; app: %s; uri: %s; err: %s") %
             app_pair.first % app_pair.second % run_res.err)};
 
         LOG_ERROR << err_desc;
-        // Do we need to set some flag for the uboot and trigger a system reboot in order to boot on a previous
-        // ostree version, hence a proper/full rollback happens???
-        ir.description += ", however " + err_desc;
-        ir.description += "\n# Apps running:\n" + getRunningAppsInfoForReport();
-        // this is a hack to distinguish between ostree install (rollback) and App start failures.
-        // data::ResultCode::Numeric::kInstallFailed - boot on a new ostree version failed (rollback at boot)
-        // data::ResultCode::Numeric::kCustomError - boot on a new version was successful but new App failed to start
-        return data::InstallationResult(run_res.imagePullFailure() ? data::ResultCode::Numeric::kDownloadFailed
-                                                                   : data::ResultCode::Numeric::kCustomError,
-                                        ir.description);
+        //Make sure not "No space or imagePullFailure"
+        if (!run_res.imagePullFailure() && !run_res.noSpace()) {
+          LOG_INFO << "(Re-)downloading app before retrying to start it again; app: " << app_pair.first;
+          const DownloadResult download_res = Download(Target::toTufTarget(target)); // download the missing apps
+          if(!download_res) {
+            LOG_ERROR << "Failed to (re-)download the missing apps; app: " << app_pair.first;
```

Comment:
Does it make sense to continue the `for` loop if an error to re-download app occurs?